### DISCUSSION TITLE
Feasibility of Using Library With Just Unique Usernames

### DISCUSSION THREAD

**Discussion by NotoriousRebel** [fastapi-users/fastapi-users#414] [Q&A]
Hello,

I recently stumbled upon this project and it looks fantastic, and I would really like to use this for a project I am working on. However, for usernames instead of the username being email-based, it's just based upon a unique username in my project. Hypothetically speaking, if I wanted to fork and tweak the library under the hood. How difficult would it be to have it so I have the option of having usernames upon email or unique usernames?

**Comment by frankie567** [comment]
Hi there,

This is not something that's supported, and I don't wish to support it for now.

However, it should indeed be quite straightforward to do it in a fork. Here is the plan:

* Add `username` property to the User models.
* Tweak `create_user` function to check username unicity.
* Add a `get_by_username` method in DB classes.
* Tweak `authenticate` method in `BaseUserDatabase` to login with a username.

I think you should have most of it working doing this.

Cheers!
